### PR TITLE
Fix docker-compose.yml

### LINKED ISSUE #1099: ./scripts/resetdb.sh appears to assume MySQL is on the default Unix socket
The `README` (perhaps) ambiguously states that MySQL should

"listen on the standard MySQL port 3306 (so mysql --host=127.0.0.1 --port=3306 connects OK)". 

This appears incorrect.

The `./scripts/resetdb.sh` appears (not an expert) to assume that `mysqld` is running on the default TCP socket and fails if, as in my case, I have a remote Cloud SQL MySQL instance proxyied to 127.0.0.1:3306.

```
./scripts/resetdb.sh
Warning: about to destroy and reset database 'test'
Are you sure? [Y/N]: Y
Resetting DB...
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2 "No such file or directory")
```

Whereas, I can happily:
```
docker run \
--interactive \
--tty \
--rm \
--net=host \
mysql mysql --host=127.0.0.1 --port=3306 --user=root
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 152
Server version: 5.6.36-google (Google)
```

I jimmied the script to force it to use the port and it works. Unfortunately my bash fu is insufficient to understand the approach in the provided script and I'm unable to provide an elegant fix:
```
mysql "${FLAGS[@]}" --host=127.0.0.1 --port=3306 -e 
```

The integration tests (`./integration/integration_tests.sh`) don't appear to make this assumption (or they assume `localhost:3306`) and they proceed correctly.

Recommend:
-- Fix the `./scripts/resetdb.sh` to either not assume the server's location or permit specifying it

### LINKED ISSUE #1164: [docker-compose] containers unable to connect to MariaDB
`docker-compose up --build` yields:

```
trillian-log-server_1  | W0529 21:19:53.108605       7 tree_storage.go:81] Failed to set strict mode on mysql db: dial tcp 172.24.190.65:3306: getsockopt: connection timed out
trillian-log-server_1  | F0529 21:19:53.109744       7 main.go:99] Failed to get storage provider: dial tcp 172.24.190.65:3306: getsockopt: connection timed out
trillian-log-server_1  | W0529 21:22:04.180614       7 tree_storage.go:81] Failed to set strict mode on mysql db: dial tcp 172.24.190.65:3306: getsockopt: connection timed out
trillian-log-server_1  | F0529 21:22:04.181499       7 main.go:99] Failed to get storage provider: dial tcp 172.24.190.65:3306: getsockopt: connection timed out
deployment_trillian-log-signer_1 exited with code 1
trillian-log-signer_1  | I0529 21:17:42.815491       7 main.go:82] **** Log Signer Starting ****
trillian-log-signer_1  | W0529 21:19:53.108579       7 tree_storage.go:81] Failed to set strict mode on mysql db: dial tcp 172.24.190.65:3306: getsockopt: connection timed out
trillian-log-signer_1  | F0529 21:19:53.109188       7 main.go:88] Failed to get storage provider: dial tcp 172.24.190.65:3306: getsockopt: connection timed out
trillian-log-signer_1  | I0529 21:19:55.425295       7 main.go:82] **** Log Signer Starting ****
trillian-log-signer_1  | W0529 21:22:06.228573       7 tree_storage.go:81] Failed to set strict mode on mysql db: dial tcp 172.24
```

and `trillian-db-seed` appears to never get a connection (through `wait-for-it`) to the database but I think it's not `wait-for-it` that's at fault as if I replace this with a long `sleep 15s` and try that way, I get errors too:
```
trillian-db-seed_1     | - Using MySQL Flags: -h mysql -u root --host localhost --port 3306
trillian-db-seed_1     | Warning: about to destroy and reset database 'test'
trillian-db-seed_1     | 
trillian-db-seed_1     | Resetting DB...
trillian-db-seed_1     | ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2 "No such file or directory")
```
IIUC it should not be attempting to connect through that socket but the port. The output suggests that it is using `localhost:3306` but this doesn't work even if I change this to `mysql` using `$DB_HOST`. 
```
trillian-db-seed_1     | ERROR 2003 (HY000): Can't connect to MySQL server on 'mysql' (110 "Connection timed out")
```

The database is ready:
```
2018-05-29 21:17:42 140010780272576 [Note] mysqld: ready for connections.
Version: '10.1.33-MariaDB-1~jessie'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  mariadb.org binary distribution
```

If I add `adminer` into the compose file, I'm able to reach the database using it:
```
  adminer:
    image: adminer
    restart: always
    ports:
      - 8080:8080
    links:
    - mysql:db
```
and then I can `localhost:8080` with `${DB_PASSWORD}`
image
In this case, it appears to have gotten as far as creating (a) `test` DB but it contains no tables; (b) `test` user.

There's something in the bowels of docker-compose and|or MySQL(MariaDB) and|or Trillian that I'm not seeing because I'm entirely unable to understand what's broken and how to fix it and would value some inspiration.
```
docker --version
Docker version 18.03.0-ce, build 0520e24
```
and
```
docker-compose --version
docker-compose version 1.21.0, build 5920eb0
```

### LINKED ISSUE #1263: Local deployment using docker-compose fails
Operating system: Ubuntu 16

Steps followed:
As mentioned here for `local deployment`:
https://github.com/google/trillian/tree/master/examples/deployment

Relevant Error on console:

```
trillian-log-server_1  | F0827 14:20:15.573541       1 main.go:100] Failed to get storage provider: dial tcp 127.0.0.1:3306: connect: connection refused
trillian-log-server_1  | F0827 14:20:16.809797       1 main.go:100] Failed to get storage provider: dial tcp 127.0.0.1:3306: connect: connection refused
trillian-log-server_1  | F0827 14:20:17.988925       1 main.go:100] Failed to get storage provider: dial tcp 127.0.0.1:3306: connect: connection refused
trillian-log-signer_1  | F0827 14:20:13.281327       1 main.go:92] Failed to get storage provider: dial tcp 127.0.0.1:3306: connect: connection refused
trillian-log-signer_1  | F0827 14:20:14.520931       1 main.go:92] Failed to get storage provider: dial tcp 127.0.0.1:3306: connect: connection refused
```

Complete log:
[output.txt](https://github.com/google/trillian/files/2324216/output.txt)

### PR DESCRIPTION
**Pull request by RJPercival**
Fixes #1263, fixes #1099 and fixes #1164 by:
- Refactoring scripts/resetdb.sh to work when the test database is being accessed from a different container.
- Passing the correct flags to Trillian.
- Adding a docker-compose step to our Travis builds, so that docker-compose.yml can't silently break in the future (at least not completely - it's only a smoke test).

### CODE REVIEW

**Comment by RJPercival** [comment]
Looks like the integration test is having some trouble with the refactored resetdb.sh script - I'll look into it.

File: .travis.yml [hunk 1]
```diff
@@ -109,6 +109,18 @@ script:
       ./trillian/integration/integration_test.sh
       cd $HOME/gopath/src/github.com/google/trillian
       HAMMER_OPTS="--operations=50" ./integration/maphammer.sh 3
+  - |
```
Comment (gdbelvin) [inline .travis.yml]: Do we have similar tests for AWS and GCP?

File: examples/deployment/docker-compose.yml [hunk 2]
```diff
@@ -3,37 +3,54 @@ services:
   mysql:
     image: mariadb:10.1
     environment:
-      - MYSQL_ROOT_PASSWORD=$DB_PASSWORD
+      - MYSQL_ROOT_PASSWORD
   trillian-db-seed:
```
Comment (gdbelvin) [inline examples/deployment/docker-compose.yml]: Should `trillian-db-seed` depend on `mysql` ?

File: examples/deployment/docker-compose.yml [hunk 3]
```diff
@@ -3,37 +3,54 @@ services:
...
     restart: always # retry while mysql is starting up
     ports:
       - "8090:8090"
       - "8091:8091"
     depends_on:
-      - mysql
-    environment:
-      - DB_USER=root
-      - DB_PASSWORD=$DB_PASSWORD
+      - trillian-db-seed
```
Comment (gdbelvin) [inline examples/deployment/docker-compose.yml]: Is `depends_on` correct?  
docker compose doesn't wait for the dependency to be healthy, only started.  Does that make a difference?

File: .travis.yml (hunk 1 above)
Comment (RJPercival) [reply .travis.yml]: We have something similar for GCP, as a consequence of Trillian's CI environment being on GCP. We don't have any tests for AWS though.

File: examples/deployment/docker-compose.yml (hunk 2 above)
Comment (RJPercival) [reply examples/deployment/docker-compose.yml]: Done.

**Comment by RJPercival** [comment]
@gdbelvin: I've addressed your comments, and added 4 commits to this branch.

**Comment by RJPercival** [comment]
I've added a couple more commits:
- Pass `$GOFLAGS` through to Docker, so Trillian is built with the appropriate flags.
- Disable the docker-compose integration test if PKCS#11 support is enabled.

Travis seems happy, so I'll merge this tomorrow unless I hear any objections.

**Comment by gdbelvin** [comment]
SGTM

**Pull request merged**